pkg/boot/linux/load_linux_arm64.go

```diff
@@ -0,0 +1,235 @@
+	}
+	if s.Size() == 0 {
+		return nil, nil, fmt.Errorf("cannot mmap zero-len file")
+	}
+	d, err := unix.Mmap(int(fd), 0, int(s.Size()), syscall.PROT_READ, syscall.MAP_PRIVATE)
+	if err != nil {
+		return nil, nil, fmt.Errorf("mmap failed: %w", err)
+	}
+
+	ummap = func() error {
+		return unix.Munmap(d)
+	}
+
+	return d, ummap, nil
+}
+
+// sanitizeFDT cleanups boot param properties from chosen node of the given FDT.
+func sanitizeFDT(fdt *dt.FDT) (*dt.Node, error) {
+	// Clear old entries in case we've already been through kexec to get
+	// to this instance of runtime.
+	chosen, _ := fdt.NodeByName("chosen")
+	if chosen == nil {
+		return nil, fmt.Errorf("no /chosen node in device tree")
+	}
+	for _, property := range []string{"linux,elfcorehdr", "linux,usable-memory-range", "kaslr-seed", "rng-seed", "linux,initrd-start", "linux,initrd-end"} {
+		chosen.RemoveProperty(property)
+	}
+
+	return chosen, nil
+}
+
+// KexecLoad loads arm64 Image, with the given ramfs and kernel cmdline.
+func KexecLoad(kernel, ramfs *os.File, cmdline string, opts KexecOptions) error {
+	var err error
+	// kmem is a struct holding kexec segments.
+	//
+	// It has routines to work with physical memory
+	// ranges.
+	var kmem *kexec.Memory
+	var kernelRange, ramfsRange, dtbRange, trampolineRange kexec.Range
+
+	fdt, err := dt.LoadFDT(opts.DTB)
+	if err != nil {
+		return fmt.Errorf("loadFDT(%s) = %v", opts.DTB, err)
+	}
+	Debug("Loaded FDT: %s", fdt)
+
+	chosen, err := sanitizeFDT(fdt)
+	if err != nil {
+		return fmt.Errorf("sanitizeFDT(%v) = %v", fdt, err)
+	}
+	Debug("FDT after sanitization: %s", fdt)
+
+	// Prepare segments.
+	kmem = &kexec.Memory{}
+	Debug("Try parsing memory map...")
+	if err := kmem.ParseMemoryMapFromFDT(fdt); err != nil {
+		return fmt.Errorf("ParseMemoryMapFromFDT(%v): %v", fdt, err)
+	}
+	Debug("Mem map: \n%+v", kmem.Phys)
+
+	// Load kernel.
+	var kernelBuf []byte
+	if opts.MmapKernel {
+		Debug("Mmapping kernel to virtual buffer...")
+		var cleanup func() error
+		kernelBuf, cleanup, err = mmap(kernel)
+		if err != nil {
+			return fmt.Errorf("mmap kernel: %v", err)
+		}
+		defer func() {
+			if err = cleanup(); err != nil {
+				Debug("Ummap kernel failed: %v", err)
+			}
+		}()
+	} else {
+		Debug("Read kernel from file ...")
+		kernelBuf, err = uio.ReadAll(kernel)
+		if err != nil {
+			return fmt.Errorf("read kernel from file: %v", err)
+		}
+	}
+
+	kImage, err := image.ParseFromBytes(kernelBuf)
+	if err != nil {
+		return fmt.Errorf("parse arm64 Image from bytes: %v", err)
+	}
+
+	Debug("Image: %+v", kImage)
+	if kernelRange, err = kmem.AddKexecSegmentExplicit(kernelBuf, uint(kImage.Header.ImageSize+kImage.Header.TextOffset), uint(kImage.Header.TextOffset), kernelAlignSize); err != nil {
+		return fmt.Errorf("add kernel segment: %v", err)
+	}
+
+	Debug("Added %d byte (size %d) kernel at %s", len(kernelBuf), kImage.Header.ImageSize, kernelRange)
+
+	var ramfsBuf []byte
+	if ramfs != nil {
+		if opts.MmapRamfs {
```

Comment:
same question here. You've got two ways to read it in; is it the case that one will work and one will not? IOW, have you seen a case where mmap always works, and ReadAll always fails, or vice versa? If both will always work, why not just pick one? If one of them will always work and the other one can sometimes fail, why not pick the one that always work? I'm really confused by this code.

Reply:
I don't know if one of them will work everywhere. Adding both ways here just to have options. I'll likely know more as we actually start rolling out either one of them to machines. We can discuss on nuances between this two reading approaches, and I probably would claim mmap will likely work everywhere (IF and only IF I have to pick one).

But, for now, I hope we have options.  IF we are going to make one of them be the de facto reading everywhere, I think we can do that by setting the the flag value, e.g. `MmapRamfs` as true everywhere.

Downside maybe the additional if/else branch adds in additional bytes in the final linked binary (if we are ever not going to use them) ?